base/inference.jl

```diff
@@ -307,8 +307,17 @@ cmp_tfunc = (x::ANY, y::ANY) -> Bool
 const _Type_name = Type.body.name
 isType(t::ANY) = isa(t,DataType) && (t::DataType).name === _Type_name
 
+_isleaf(t::ANY) = isleaftype(t) || t === Union{}
+
 # true if Type is inlineable as constant (is a singleton)
-isconstType(t::ANY) = isType(t) && isleaftype(t.parameters[1])
+function isconstType(t::ANY)
+    if isType(t)
+        p = t.parameters[1]
+        return (_isleaf(p) ||
+                (isa(p,DataType) && p <: Tuple && all(_isleaf, p.parameters)))
+    end
+    return false
+end
```

Comment:
should be able to replace this part of the commit with 1f66a6672f6a313f9393b20f16456dfeba938fe6